### PR TITLE
[pytorch][PR] introduce WeakHashRef

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115216

We would like weak dictionaries that have `torch.ScriptObject` keys. Similar to tensors, we need to override the behavior of the ref to dot he right thing under comparison.

This change also makes it so that WeakIdKeyDictionary works with a pluggable ref_type.

Differential Revision: [D51828205](https://our.internmc.facebook.com/intern/diff/D51828205/)

### CODE REVIEW

**Comment by suo** [comment]
ignore the lint failure; https://github.com/pytorch/pytorch/pull/115233 should move that check to use a different thing

File: test/test_weak.py [hunk 1]
```diff
@@ -562,5 +570,340 @@ def test_pop(self):
...
+                "//caffe2/test/cpp/jit:test_custom_class_registrations"
+            )
+        else:
+            lib_file_path = find_library_location("libtorchbind_test.so")
+            if IS_WINDOWS:
+                lib_file_path = find_library_location("torchbind_test.dll")
+            torch.ops.load_library(str(lib_file_path))
+
+        unittest.TestCase.__init__(self, *args, **kw)
+        self.reference = self._reference().copy()
```
Comment (albanD) [inline test/test_weak.py:610]: This will be shared by all tests. Shouldn't this be in setUp() ?

File: test/test_weak.py [hunk 2]
```diff
@@ -562,5 +570,340 @@ def test_pop(self):
...
+
+        # A (key, value) pair in the mapping
+        key, value = self.reference.popitem()
+        self.inmapping = {key: value}
+        self.reference[key] = value
+
+    def test_read(self):
+        # Test for read only operations on mapping
+        p = self._empty_mapping()
+        p1 = dict(p)  # workaround for singleton objects
```
Comment (albanD) [inline test/test_weak.py:624]: What does this mean?

**Review by albanD — COMMENTED** [review]
Tests are a bit confusing haha

File: docs/source/conf.py [hunk 3]
```diff
@@ -3487,6 +3487,7 @@
     "WeakIdKeyDictionary",
     "WeakIdRef",
     "WeakTensorKeyDictionary",
+    "WeakScriptObjectRef",
```
[the remark this answers is not in the document]
Comment (suo) [reply docs/source/conf.py:3490]: I'll make WeakScriptObjectRef hidden per discussion

File: torch/utils/weak.py [hunk 4]
```diff
@@ -80,13 +80,42 @@ def __eq__(self, other):
             return a is b
         return self is other
 
+# ScriptObjects are wrappers over IValues which contain a torchbind class. For
+# the purpose of this class, we treat two different ScriptObjects that wrap the
+# same IValue as the equal.
+class WeakScriptObjectRef(weakref.ref):
+    __slots__ = ['_id']
+
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/utils/weak.py:88]: oh I see, you wanted to make it more generic

File: test/test_weak.py (hunk 1 above)
Comment (suo) [reply test/test_weak.py:610]: this is part of the code copied from cpython; I'd prefer not to change unless necessary lol

File: torch/utils/weak.py [hunk 5]
```diff
@@ -4,14 +4,14 @@
...
-from torch import Tensor
+from torch import Tensor, ScriptObject
 import collections.abc as _collections_abc
 
 
 WeakRef = ref
 
 
-__all__ = ['TensorWeakRef', 'WeakIdRef', 'WeakIdKeyDictionary', 'WeakTensorKeyDictionary']
+__all__ = ['TensorWeakRef', 'WeakIdRef', 'WeakIdKeyDictionary', 'WeakTensorKeyDictionary', '_WeakHashRef']
```
Comment (albanD) [inline torch/utils/weak.py:14]: Don't put it here, this should contain only the public APIs.

**Review by albanD — APPROVED** [review]
Small details on the public API but good to go once the CI is green!

**Pull request closed without merging**